File: SvgEditor.cs
// SvgEditor.cs
// Версия на C# с использованием System.Xml, LINQ, record types

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Xml;

public record LayerInfo(int Index, string Id, Dictionary<string, string> Attributes);

public class SvgEditor
{
    private readonly string _filename;
    private XmlDocument _doc;
    private XmlElement _svgRoot;
    private List<XmlElement> _layers = new();

    public SvgEditor(string filename)
    {
        _filename = filename;
        _doc = new XmlDocument();
        _doc.Load(filename);
        _svgRoot = _doc.DocumentElement;
        LoadLayers();
    }

    private void LoadLayers()
    {
        _layers = _svgRoot.ChildNodes
            .OfType<XmlElement>()
            .Where(e => e.Name == "g" || e.HasAttribute("data-layer"))
            .ToList();
    }

    public void ListLayers()
    {
        if (!_layers.Any())
        {
            Console.WriteLine("Слои не найдены.");
            return;
        }
        for (int i = 0; i < _layers.Count; i++)
        {
            var el = _layers[i];
            string id = el.GetAttribute("id");
            if (string.IsNullOrEmpty(id)) id = el.GetAttribute("data-layer");
            string attrs = string.Join(" ", el.Attributes.Cast<XmlAttribute>().Select(a => $"{a.Name}=\"{a.Value}\""));
            Console.WriteLine($"[{i}] {(string.IsNullOrEmpty(id) ? "без id" : id)} : {attrs}");
        }
    }

    public void AddLayer(string id, Dictionary<string, string> extraAttrs)
    {
        var newGroup = _doc.CreateElement("g");
        if (!string.IsNullOrEmpty(id)) newGroup.SetAttribute("id", id);
        foreach (var kv in extraAttrs)
            newGroup.SetAttribute(kv.Key, kv.Value);
        _svgRoot.AppendChild(newGroup);
        _layers.Add(newGroup);
        Console.WriteLine($"Слой '{id}' добавлен.");
    }

    public void RemoveLayer(int index)
    {
        if (index < 0 || index >= _layers.Count) throw new IndexOutOfRangeException("Индекс вне диапазона");
        var el = _layers[index];
        _svgRoot.RemoveChild(el);
        _layers.RemoveAt(index);
        Console.WriteLine($"Слой {index} удалён.");
    }

    public void MoveLayer(int index, string direction)
    {
        if (index < 0 || index >= _layers.Count) throw new IndexOutOfRangeException("Индекс вне диапазона");
        int newIdx = direction == "up" ? index - 1 : index + 1;
        if (newIdx < 0 || newIdx >= _layers.Count) throw new InvalidOperationException("Невозможно переместить");
        var el1 = _layers[index];
        var el2 = _layers[newIdx];
        _svgRoot.RemoveChild(el1);
        _svgRoot.RemoveChild(el2);
        if (index < newIdx)
        {
            _svgRoot.InsertBefore(el2, el1);
            _svgRoot.InsertBefore(el1, el2);
        }
        else
        {
            _svgRoot.InsertBefore(el1, el2);
            _svgRoot.InsertBefore(el2, el1);
        }
        (_layers[index], _layers[newIdx]) = (_layers[newIdx], _layers[index]);
        Console.WriteLine($"Слой {index} перемещён {direction}.");
    }

    public void EditLayer(int index, string attr, string value)
    {
        if (index < 0 || index >= _layers.Count) throw new IndexOutOfRangeException("Слой не найден");
        var el = _layers[index];
        el.SetAttribute(attr, value);
        Console.WriteLine($"Атрибут '{attr}' слоя {index} установлен в '{value}'.");
    }

    public void Save(string output)
    {
        string outFile = string.IsNullOrEmpty(output) ? _filename : output;
        _doc.Save(outFile);
        Console.WriteLine($"Сохранено в {outFile}");
    }

    public static void Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Использование: dotnet run -- <input.svg> <команда> [параметры]");
            return;
        }
        string input = args[0];
        string command = args[1];
        var editor = new SvgEditor(input);

        try
        {
            switch (command)
            {
                case "list":
                    editor.ListLayers();
                    break;
                case "add-layer":
                {
                    string id = null;
                    var extra = new Dictionary<string, string>();
                    for (int i = 2; i < args.Length; i++)
                    {
                        if (args[i] == "--id" && i + 1 < args.Length)
                            id = args[++i];
                        else if (args[i] == "--attr" && i + 1 < args.Length)
                        {
                            var kv = args[++i].Split('=');
                            if (kv.Length == 2) extra[kv[0]] = kv[1];
                        }
                    }
                    editor.AddLayer(id, extra);
                    editor.Save(null);
                    break;
                }
                case "remove-layer":
                {
                    int index = -1;
                    for (int i = 2; i < args.Length; i++)
                    {
                        if (args[i] == "--index" && i + 1 < args.Length)
                            index = int.Parse(args[++i]);
                    }
                    if (index < 0) throw new Exception("Не указан --index");
                    editor.RemoveLayer(index);
                    editor.Save(null);
                    break;
                }
                case "move-layer":
                {
                    int index = -1;
                    string dir = "";
                    for (int i = 2; i < args.Length; i++)
                    {
                        if (args[i] == "--index" && i + 1 < args.Length)
                            index = int.Parse(args[++i]);
                        else if (args[i] == "--direction" && i + 1 < args.Length)
                            dir = args[++i];
                    }
                    if (index < 0 || string.IsNullOrEmpty(dir)) throw new Exception("Не указаны --index и --direction");
                    editor.MoveLayer(index, dir);
                    editor.Save(null);
                    break;
                }
                case "edit-layer":
                {
                    int index = -1;
                    string attr = "", value = "";
                    for (int i = 2; i < args.Length; i++)
                    {
                        if (args[i] == "--index" && i + 1 < args.Length)
                            index = int.Parse(args[++i]);
                        else if (args[i] == "--attr" && i + 1 < args.Length)
                            attr = args[++i];
                        else if (args[i] == "--value" && i + 1 < args.Length)
                            value = args[++i];
                    }
                    if (index < 0 || string.IsNullOrEmpty(attr) || string.IsNullOrEmpty(value))
                        throw new Exception("Не указаны --index, --attr, --value");
                    editor.EditLayer(index, attr, value);
                    editor.Save(null);
                    break;
                }
                case "save":
                {
                    string output = null;
                    for (int i = 2; i < args.Length; i++)
                    {
                        if (args[i] == "--output" && i + 1 < args.Length)
                            output = args[++i];
                    }
                    editor.Save(output);
                    break;
                }
                default:
                    Console.WriteLine($"Неизвестная команда: {command}");
                    break;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Ошибка: {ex.Message}");
        }
    }
}

File: svg_editor.php
<?php
// svg_editor.php
// Версия на PHP 8 с DOMDocument, атрибутами, генераторами

class SvgEditor {
    private string $filename;
    private DOMDocument $doc;
    private DOMElement $svgRoot;
    private array $layers = [];

    public function __construct(string $filename) {
        $this->filename = $filename;
        $this->doc = new DOMDocument();
        $this->doc->load($filename);
        $this->svgRoot = $this->doc->documentElement;
        $this->loadLayers();
    }

    private function loadLayers(): void {
        $this->layers = [];
        foreach ($this->svgRoot->childNodes as $node) {
            if ($node->nodeType === XML_ELEMENT_NODE) {
                $el = $node;
                if ($el->tagName === 'g' || $el->hasAttribute('data-layer')) {
                    $this->layers[] = $el;
                }
            }
        }
    }

    public function listLayers(): void {
        if (empty($this->layers)) {
            echo "Слои не найдены.\n";
            return;
        }
        foreach ($this->layers as $i => $el) {
            $id = $el->getAttribute('id') ?: $el->getAttribute('data-layer') ?: 'без id';
            $attrs = [];
            foreach ($el->attributes as $attr) {
                $attrs[] = $attr->name . '="' . $attr->value . '"';
            }
            echo sprintf("[%d] %s : %s\n", $i, $id, implode(' ', $attrs));
        }
    }

    public function addLayer(string $id, array $extraAttrs): void {
        $newGroup = $this->doc->createElement('g');
        if ($id) $newGroup->setAttribute('id', $id);
        foreach ($extraAttrs as $k => $v) {
            $newGroup->setAttribute($k, $v);
        }
        $this->svgRoot->appendChild($newGroup);
        $this->layers[] = $newGroup;
        echo "Слой '$id' добавлен.\n";
    }

    public function removeLayer(int $index): void {
        if (!isset($this->layers[$index])) {
            throw new Exception("Индекс вне диапазона");
        }
        $el = $this->layers[$index];
        $this->svgRoot->removeChild($el);
        array_splice($this->layers, $index, 1);
        echo "Слой $index удалён.\n";
    }

    public function moveLayer(int $index, string $direction): void {
        if (!isset($this->layers[$index])) {
            throw new Exception("Индекс вне диапазона");
        }
        $newIdx = $direction === 'up' ? $index - 1 : $index + 1;
        if ($newIdx < 0 || $newIdx >= count($this->layers)) {
            throw new Exception("Невозможно переместить");
        }
        $el1 = $this->layers[$index];
        $el2 = $this->layers[$newIdx];
        // Перемещаем в DOM
        if ($index < $newIdx) {
            $this->svgRoot->insertBefore($el2, $el1);
            $this->svgRoot->insertBefore($el1, $el2);
        } else {
            $this->svgRoot->insertBefore($el1, $el2);
            $this->svgRoot->insertBefore($el2, $el1);
        }
        // меняем местами в массиве
        $this->layers[$index] = $el2;
        $this->layers[$newIdx] = $el1;
        echo "Слой $index перемещён $direction.\n";
    }

    public function editLayer(int $index, string $attr, string $value): void {
        if (!isset($this->layers[$index])) {
            throw new Exception("Слой не найден");
        }
        $this->layers[$index]->setAttribute($attr, $value);
        echo "Атрибут '$attr' слоя $index установлен в '$value'.\n";
    }

    public function save(?string $output = null): void {
        $out = $output ?? $this->filename;
        $this->doc->save($out);
        echo "Сохранено в $out\n";
    }
}

// Парсинг аргументов
if ($argc < 3) {
    echo "Использование: php svg_editor.php <input.svg> <команда> [параметры]\n";
    exit(1);
}

$input = $argv[1];
$command = $argv[2];

$editor = new SvgEditor($input);

try {
    switch ($command) {
        case 'list':
            $editor->listLayers();
            break;
        case 'add-layer':
            $id = null;
            $extra = [];
            for ($i = 3; $i < $argc; $i++) {
                if ($argv[$i] === '--id' && isset($argv[$i+1])) {
                    $id = $argv[++$i];
                } elseif ($argv[$i] === '--attr' && isset($argv[$i+1])) {
                    $kv = explode('=', $argv[++$i], 2);
                    if (count($kv) === 2) $extra[$kv[0]] = $kv[1];
                }
            }
            $editor->addLayer($id, $extra);
            $editor->save(null);
            break;
        case 'remove-layer':
            $index = -1;
            for ($i = 3; $i < $argc; $i++) {
                if ($argv[$i] === '--index' && isset($argv[$i+1])) {
                    $index = (int)$argv[++$i];
                }
            }
            if ($index < 0) throw new Exception("Не указан --index");
            $editor->removeLayer($index);
            $editor->save(null);
            break;
        case 'move-layer':
            $index = -1; $dir = '';
            for ($i = 3; $i < $argc; $i++) {
                if ($argv[$i] === '--index' && isset($argv[$i+1])) {
                    $index = (int)$argv[++$i];
                } elseif ($argv[$i] === '--direction' && isset($argv[$i+1])) {
                    $dir = $argv[++$i];
                }
            }
            if ($index < 0 || empty($dir)) throw new Exception("Не указаны --index и --direction");
            $editor->moveLayer($index, $dir);
            $editor->save(null);
            break;
        case 'edit-layer':
            $index = -1; $attr = ''; $value = '';
            for ($i = 3; $i < $argc; $i++) {
                if ($argv[$i] === '--index' && isset($argv[$i+1])) {
                    $index = (int)$argv[++$i];
                } elseif ($argv[$i] === '--attr' && isset($argv[$i+1])) {
                    $attr = $argv[++$i];
                } elseif ($argv[$i] === '--value' && isset($argv[$i+1])) {
                    $value = $argv[++$i];
                }
            }
            if ($index < 0 || empty($attr) || empty($value)) {
                throw new Exception("Не указаны --index, --attr, --value");
            }
            $editor->editLayer($index, $attr, $value);
            $editor->save(null);
            break;
        case 'save':
            $output = null;
            for ($i = 3; $i < $argc; $i++) {
                if ($argv[$i] === '--output' && isset($argv[$i+1])) {
                    $output = $argv[++$i];
                }
            }
            $editor->save($output);
            break;
        default:
            echo "Неизвестная команда: $command\n";
    }
} catch (Exception $e) {
    echo "Ошибка: " . $e->getMessage() . "\n";
    exit(1);
}
